scope_shared/scope/documents/document_set.py

```diff
@@ -0,0 +1,399 @@
+
+        # Obtain a list of items
+        try:
+            other_items: List = list(iter(other))
+        except TypeError:
+            return False
+
+        # If the lists are not the same length, we cannot be equal
+        if len(self) != len(other_items):
+            return False
+
+        # Every item in our set must appear exactly once in the other items
+        for item_current in self:
+            try:
+                other_items.remove(item_current)
+            except ValueError:
+                return False
+
+        return True
+
+    def __len__(self) -> int:
+        """
+        Number of documents in the set.
+        """
+        return len(self.documents)
+
+    def filter_match(
+        self,
+        match_type: Optional[str] = None,
+        match_deleted: Optional[bool] = None,
+        match_datetime_at: Optional[datetime] = None,
+        match_values: Optional[Dict[str, str]] = None,
+    ) -> DocumentSet:
+        """
+        Keep only documents that match all provided parameters.
+        """
+
+        retained_documents = []
+        for document_current in self:
+            matches: bool = self._match_document(
+                document=document_current,
+                match_type=match_type,
+                match_deleted=match_deleted,
+                match_datetime_at=match_datetime_at,
+                match_values=match_values,
+            )
+
+            if matches:
+                retained_documents.append(document_current)
+
+        return DocumentSet(documents=retained_documents)
+
+    def is_unique(self) -> bool:
+        """
+        Whether this set contains exactly one document.
+        """
+        return len(self) == 1
+
+    def __iter__(self) -> Iterable[dict]:
+        """
+        Iterate over contained documents.
+        """
+        return iter(self.documents)
+
+    def _match_document(
+        self,
+        document: dict,
+        match_type: Optional[str],
+        match_deleted: Optional[bool],
+        match_datetime_at: Optional[datetime],
+        match_values: Optional[Dict[str, str]],
+    ) -> bool:
+        tested: bool = False
+        matches: bool = True
+
+        if matches and match_type:
+            tested = True
+            matches = matches and document["_type"] == match_type
+
+        if matches and match_deleted is not None:
+            tested = True
+            matches = matches and match_deleted == document.get("_deleted", False)
+
+        if matches and match_datetime_at is not None:
+            tested = True
+
+            if match_datetime_at < datetime_from_document(document=document):
+                # A document created after our match time cannot match
+                matches = matches and False
+            else:
+                # A document created at or before our match time
+                # matches if no revision replaces it before the match time
+                revisions = self.revisions[document_key(document=document)]
```

Comment:
Question for you, how do I get a set of _all_ documents of a particular `_type` with _all_ the revisions before a specific datetime?

Reply:
Not currently implemented, but it would be a `match_datetime_before` parameter.

Do you have a current use case?